### PR TITLE
feat(TV 2 Play): add activity

### PR DESCRIPTION
**Pull request by NoOneNook**
## Description
Adds a new TV 2 Play activity that shows what the user is watching on TV 2 Play, including series/movie information, season/episode details, playback state, artwork, and timestamps when available.

## Changes in the latest update
- Added Privacy Mode. While watching, it hides the exact title and instead shows a generic media type such as series, movie, or live TV. Normal browsing presence is not overridden by Privacy Mode.
- Replaced the manual Norwegian/English text table with PreMiD's built-in `presence.getStrings()` localization using `general.*` keys.
- Updated the metadata description to the official TV 2 Play service description.
- Updated the URL regex to the repository reviewer's requested format.
- Combined async settings/localization work with `Promise.all`.
- Muted autoplay previews on series/movie detail pages are ignored, so simply opening a title page does not incorrectly appear as active playback.
- Short previews with audible audio are detected as trailers.
- Hidden/detached source-backed video elements are ignored after SPA navigation.
- Episode-title detection accepts multiple marker formats such as `S1E2`, `S1 E2`, `Sesong/Season N`, and `Episode/Ep. N`.
- Live detection is restricted to the actual `/direkte-tv` route.

### Muted autoplay previews are ignored
TV 2 Play can automatically play a muted preview on series/movie detail pages. These previews are now ignored so the activity remains in the normal browsing state.

TV 2 Play browsing while muted autoplay preview is ignored

### Audible trailers are detected
If a short preview is actually playing with audio, it is treated as a trailer rather than regular movie/episode playback.

TV 2 Play trailer detection

### Privacy Mode while watching
Privacy Mode hides the exact title and detailed playback information while still indicating the general type of content being watched.

TV 2 Play Privacy Mode

## Acknowledgements
- [x] I read the [Activity Guidelines](https://github.com/PreMiD/Activities/blob/main/.github/CONTRIBUTING.md)
- [x] I linted the code by running `npm run lint`
- [x] The PR title follows the repo's [commit conventions](https://github.com/PreMiD/Activities/blob/main/.github/COMMIT_CONVENTION.md)

## Validation
- `npx eslint "websites/T/TV 2 Play" --no-cache` passes locally
- `npx pmd build "TV 2 Play" --validate` passes locally
- The updated activity was tested locally in PreMiD development mode
- Browsing, regular playback, Privacy Mode, muted autoplay previews, audible trailer detection, and normal episode playback were verified locally

### Playback
TV 2 Play playback presence

### Browsing
TV 2 Play browsing presence

### CODE REVIEW

File: websites/T/TV 2 Play/presence.ts [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+
+function isLiveVideo(video: HTMLVideoElement): boolean {
+  if (!Number.isFinite(video.duration) || video.duration === Infinity)
+    return true
+
+  const path = document.location.pathname.toLowerCase()
+
+  return (
+    path.includes('/direkte')
+    || path.includes('/live')
```
[the remark this answers is not in the document]
Comment (NoOneNook) [reply websites/T/TV 2 Play/presence.ts]: Fixed in ac5eb24c2: live detection is now limited to the actual `/direkte-tv` route (or its subpaths), while infinite-duration media is still treated as live.

File: websites/T/TV 2 Play/presence.ts [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+  )
+}
+
+function largestVideo(root: ParentNode = document): HTMLVideoElement | null {
+  return [...root.querySelectorAll('video')]
+    .map(video => ({
+      video,
+      rect: video.getBoundingClientRect(),
+    }))
+    .filter(({ rect }) => rect.width > 300 && rect.height > 150)
```
[the remark this answers is not in the document]
Comment (NoOneNook) [reply websites/T/TV 2 Play/presence.ts]: Fixed in ac5eb24c2: the fixed 300×150 rendered-size threshold was removed. Video selection now uses loaded/source state, prioritizes active playback, and only uses rendered area as a tie-breaker.

File: websites/T/TV 2 Play/presence.ts [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+    && Number(style.opacity) > 0.05
+  )
+}
+
+function largestVideo(root: ParentNode = document): HTMLVideoElement | null {
+  return [...root.querySelectorAll('video')]
+    .filter(video =>
+      video.readyState > 0
+      && Boolean(video.currentSrc || video.src || video.srcObject),
+    )
```
[the remark this answers is not in the document]
Comment (NoOneNook) [reply websites/T/TV 2 Play/presence.ts:108]: Fixed in 4271aff6a: video selection now requires the video to still be connected and visible, while keeping the previous fixed 300×150 threshold removed. Muted autoplay previews on title pages are also ignored.

File: websites/T/TV 2 Play/presence.ts [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+function readEpisodeTitle(
+  player: Element | null,
+  season?: string,
+  episode?: string,
+): string {
+  if (!player || !season || !episode)
+    return ''
+
+  const codeRegex
+    = new RegExp(`^S${season}E${episode}$`, 'i')
```
[the remark this answers is not in the document]
Comment (NoOneNook) [reply websites/T/TV 2 Play/presence.ts]: Fixed in 4271aff6a: episode-title extraction now accepts `S1E2`, `S1 E2`, `Sesong/Season N`, and `Episode/Ep. N` marker formats consistently with the parser.

File: websites/T/TV 2 Play/metadata.json [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  "$schema": "https://schemas.premid.app/metadata/1.17",
+  "apiVersion": 1,
+  "author": {
+    "name": "noonenook",
+    "id": "437695615095275520"
+  },
+  "service": "TV 2 Play",
+  "description": {
+    "en": "Shows what you are watching on TV 2 Play.",
+    "no": "Viser hva du ser på TV 2 Play."
```
Comment (skullysmods) [inline websites/T/TV 2 Play/metadata.json]: Use official description from service

File: websites/T/TV 2 Play/metadata.json [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  },
+  "url": "play.tv2.no",
+  "regExp": "^https?://play[.]tv2[.]no(?:/|$)",
```
Comment (skullysmods) [inline websites/T/TV 2 Play/metadata.json]:
```suggestion
  "regExp": "^https?[:][/][/]play[.]tv2[.]no[/]",
```

File: websites/T/TV 2 Play/presence.ts [hunk 7]
```diff
@@ -0,0 +1,462 @@
...
+
+  return path === '/direkte-tv' || path.startsWith('/direkte-tv/')
+}
+
+presence.on('UpdateData', async () => {
+  const language
+    = await presence.getSetting<number>('language')
+
+  const privacyMode
+    = await presence.getSetting<boolean>('privacyMode')
```
Comment (skullysmods) [inline websites/T/TV 2 Play/presence.ts]: Combine using Promise.all

File: websites/T/TV 2 Play/presence.ts [hunk 8]
```diff
@@ -0,0 +1,462 @@
...
+        pause: 'Pause',
+        button: 'Se på TV 2 Play',
+        season: 'Sesong',
+        episode: 'Episode',
+        privacy: 'Privat modus',
+        privateSeries: 'Ser på en serie',
+        privateMovie: 'Ser på en film',
+        privateLive: 'Ser på direktesendt TV',
+        trailer: 'Ser på en trailer',
+      }
```
Comment (skullysmods) [inline websites/T/TV 2 Play/presence.ts]: Use PreMiD's built-in localization feature instead

**Review by skullysmods — CHANGES_REQUESTED** [review]

File: websites/T/TV 2 Play/metadata.json (hunk 5 above)
Comment (NoOneNook) [reply websites/T/TV 2 Play/metadata.json]: Addressed in 552ed0e03: the metadata now uses the official TV 2 Play service description in English and Norwegian.

File: websites/T/TV 2 Play/metadata.json (hunk 6 above)
Comment (NoOneNook) [reply websites/T/TV 2 Play/metadata.json]: Applied in 552ed0e03 exactly as suggested: `^https?[:][/][/]play[.]tv2[.]no[/]`.

File: websites/T/TV 2 Play/presence.ts (hunk 7 above)
Comment (NoOneNook) [reply websites/T/TV 2 Play/presence.ts]: Addressed in 552ed0e03: the async settings/localization work at the top of `UpdateData` is now combined with `Promise.all`.

File: websites/T/TV 2 Play/presence.ts (hunk 8 above)
Comment (NoOneNook) [reply websites/T/TV 2 Play/presence.ts]: Addressed in 552ed0e03: the manual Norwegian/English string table was removed and the activity now uses PreMiD's built-in `presence.getStrings()` localization with `general.*` keys. The custom language dropdown was also removed after local testing, so there is no duplicate/manual language selector.

File: websites/T/TV 2 Play/presence.ts [hunk 9]
```diff
@@ -0,0 +1,433 @@
...
+    && style.display !== 'none'
+    && style.visibility !== 'hidden'
+    && Number(style.opacity) > 0.05
+  )
+}
+
+function isTitlePage(): boolean {
+  const path = document.location.pathname.toLowerCase()
+
+  return /^\/(?:serie|serier|film|filmer)\/[^/]+\/?$/.test(path)
```
[the remark this answers is not in the document]
Comment (NoOneNook) [reply websites/T/TV 2 Play/presence.ts]: Fixed in 39fcc7cba: `/programmer/<category>/<slug>` pages are now included in title-page detection, so muted autoplay previews on those pages are ignored as well.

File: websites/T/TV 2 Play/presence.ts [hunk 10]
```diff
@@ -0,0 +1,433 @@
...
+        presenceData.state = episodeTitle
+          ? `${strings.season} ${season} • ${strings.episode} ${episode} • ${episodeTitle}`
+          : `${strings.season} ${season} • ${strings.episode} ${episode}`
+      }
+      else if (live) {
+        presenceData.state = strings.live
+      }
+      else {
+        presenceData.state
+          = video.paused ? strings.paused : strings.watching
```
[the remark this answers is not in the document]
Comment (NoOneNook) [reply websites/T/TV 2 Play/presence.ts]: Fixed in 39fcc7cba: regular non-live/non-episodic playback now uses the standalone `general.playing` string instead of the `general.watching` prefix, avoiding a dangling `Watching:` state.

File: websites/T/TV 2 Play/presence.ts [hunk 11]
```diff
@@ -0,0 +1,433 @@
...
+    const trailer = isAudibleTrailer(video)
+
+    presenceData.type = ActivityType.Watching
+
+    if (privacyMode) {
+      presenceData.details = live
+        ? strings.watchingLive
+        : season && episode
+          ? strings.watchingSeries
+          : strings.watchingMovie
```
[the remark this answers is not in the document]
Comment (NoOneNook) [reply websites/T/TV 2 Play/presence.ts]: Fixed in 39fcc7cba: Privacy Mode now also uses the current title-page route to classify series/program trailers as series even when the trailer itself has no season/episode marker.

File: websites/T/TV 2 Play/presence.ts [hunk 12]
```diff
@@ -0,0 +1,444 @@
...
+}
+
+function isMutedAutoplayPreview(video: HTMLVideoElement): boolean {
+  if (!isTitlePage())
+    return false
+
+  const muted
+    = video.muted
+      || video.defaultMuted
+      || video.volume === 0
```
[the remark this answers is not in the document]
Comment (NoOneNook) [reply websites/T/TV 2 Play/presence.ts:81]: Fixed in a881c16de: preview audibility now uses the current `video.muted` and `video.volume` state only, so a preview that started muted is detected correctly after the user unmutes it.

File: websites/T/TV 2 Play/metadata.json [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+  ],
+  "settings": [
+    {
+      "id": "privacyMode",
+      "title": "Privacy Mode",
+      "icon": "fas fa-user-secret",
+      "value": false,
+      "description": "Hide the exact title and episode information while watching."
+    }
+  ]
```
Comment (skullysmods) [inline websites/T/TV 2 Play/metadata.json:38]:
```suggestion
  "settings": [
    {
      "id": "lang",
      "multiLanguage": true
    },
    {
      "id": "privacyMode",
      "title": "Privacy Mode",
      "icon": "fas fa-user-secret",
      "value": false,
      "description": "Hide the exact title and episode information while watching."
    }
  ]
```
Required for localization

**Review by skullysmods — CHANGES_REQUESTED** [review]

File: websites/T/TV 2 Play/metadata.json (hunk 13 above)
Comment (NoOneNook) [reply websites/T/TV 2 Play/metadata.json:38]: Applied in 61b255705: added the required `lang` setting with `multiLanguage: true` before Privacy Mode. Both Validate and ESLint are green on the latest commit.

File: websites/T/TV 2 Play/presence.ts [hunk 14]
```diff
@@ -0,0 +1,442 @@
...
+      video.isConnected
+      && isVisible(video)
+      && !isMutedAutoplayPreview(video)
+      && video.readyState > 0
+      && Boolean(video.currentSrc || video.src || video.srcObject),
+    )
+    .map(video => ({
+      video,
+      rect: video.getBoundingClientRect(),
+      active: !video.paused && !video.ended,
```
[the remark this answers is not in the document]
Comment (NoOneNook) [reply websites/T/TV 2 Play/presence.ts:112]: Fixed in 114e3fe50: ended video elements are now excluded directly in `largestVideo()`, so completed playback no longer remains selected as active watching. Both Validate and ESLint are green.

**Review by skullysmods — APPROVED** [review]

**Review by Bas950 — APPROVED** [review]
